src/Emogrifier/CssConcatenator.php

```diff
@@ -0,0 +1,122 @@
+ *  $css = $concatenator->getCss();
+ *
+ * `$css` (if unminified) would contain the following CSS:
+ * ` body {
+ * `   color: blue;
+ * `   font-size: 16px;
+ * ` }
+ * ` p, ul, ol {
+ * `   margin: 1em 0;
+ * ` }
+ * ` @media screen and (max-width: 400px) {
+ * `   body {
+ * `     font-size: 14px;
+ * `   }
+ * `   ul, ol {
+ * `     margin: 0.75em 0;
+ * `   }
+ * ` }
+ *
+ * @author Jake Hotson <jake.github@qzdesign.co.uk>
+ */
+class CssConcatenator
+{
+    /**
+     * Array of media rules in order.  Each element is an object with the following properties:
+     * - string `media` - The media query string, e.g. "@media screen and (max-width:639px)", or an empty string for
+     *   rules not within a media query block;
+     * - stdClass[] `ruleBlocks` - Array of rule blocks in order, where each element is an object with the following
+     *   properties:
+     *   - int[] `selectorsAsKeys` - Array whose keys are selectors for the rule block (values are of no significance);
+     *   - string `declarationsBlock` - The property declarations, e.g. "margin-top: 0.5em; padding: 0".
+     *
+     * @var stdClass[]
+     */
+    private $mediaRules = [];
+
+    /**
+     * Appends a declaration block to the CSS.
+     *
+     * @param string[] $selectors Array of selectors for the rule, e.g. ["ul", "ol", "p:first-child"].
+     * @param string $declarationsBlock The property declarations, e.g. "margin-top: 0.5em; padding: 0".
+     * @param string $media The media query for the rule, e.g. "@media screen and (max-width:639px)",
+     *                      or an empty string if none.
+     */
+    public function append(array $selectors, $declarationsBlock, $media = '')
+    {
+        $lastMediaRule = end($this->mediaRules);
+        if ($lastMediaRule !== false && $media === $lastMediaRule->media) {
+            $mediaRule = $lastMediaRule;
+        } else {
+            $mediaRule = (object)[
+                'media' => $media,
+                'ruleBlocks' => [],
+            ];
+            $this->mediaRules[] = $mediaRule;
+        }
+
+        $lastRuleBlock = end($mediaRule->ruleBlocks);
+        $selectorsAsKeys = array_flip($selectors);
+        if ($lastRuleBlock !== false && $declarationsBlock === $lastRuleBlock->declarationsBlock) {
+            $lastRuleBlock->selectorsAsKeys += $selectorsAsKeys;
+        } elseif ($lastRuleBlock !== false
+            && $this->hasEquivalentSelectors($selectorsAsKeys, $lastRuleBlock->selectorsAsKeys)
+        ) {
+            $lastRuleBlock->declarationsBlock
+                = rtrim(rtrim($lastRuleBlock->declarationsBlock), ';') . ';' . $declarationsBlock;
+        } else {
+            $mediaRule->ruleBlocks[] = (object)compact('selectorsAsKeys', 'declarationsBlock');
+        }
+    }
+
+    /**
+     * @return string
+     */
+    public function getCss()
+    {
+        $css = '';
+        foreach ($this->mediaRules as $mediaRule) {
```

Comment:
Please let's simplify this by moving parts of the code to a separate, well-named method, e.g. `buildCssForMediaRule($mediaRule)`. This will make the intention of the code more clear to the reader.

Reply:
Moved the body if each `for` loop into separate methods - which are now be passed to `array_map` avoiding the loop constructs altogether.